asvdb/asvdb.py

```diff
@@ -925,49 +943,117 @@ def __createLockfile(self, dirPath):
+            debugCounter = 0
 
+            while otherLockfileTimes:
+                debugCounter += 1
+                if self.debugPrint:
+                    print(f"This lock file will be {thisLockfile} but other "
+                          f"locks present: {otherLocks}, waiting to try to "
+                          "lock again...")
+                time.sleep(2)
+                otherLockfileTimes = self.__updateOtherLockfileTimes()
+                if debugCounter >= 10:
+                    raise Exception("Aborted due to possible infinite loop")
 
-    def __releaseS3Lock(self, url):
-        # FIXME: write this
-        raise NotImplementedError
+            # All clear, create lock
+            if self.debugPrint:
+                print(f"All clear, setting lock {thisLockfile}")
+            self.s3Resource.Object(self.bucketName, thisLockfile).put()
+
+            # Check for a race condition where another lock could have been created
+            # while creating the lock for this instance.
+            otherLockfileTimes = self.__updateS3LockfileTimes()
+            
+            if otherLockfileTimes:
+                self.__releaseS3Lock()
+                randTime = (int(10 * random.random()) + 2) + random.random()
+                if self.debugPrint:
+                    print(f"Collision - waiting {randTime} seconds before "
+                          "trying to lock again.")
+                time.sleep(randTime)
+            else:
+                break
+
+            i += 1
+
+
+            
+    def __updateS3LockfileTimes(self):
+        # Find lockfiles in S3 Bucket
+        response = self.s3Resource.list_objects_v2(
+            Bucket=self.bucketName,
+            Prefix=self.lockfilePrefix,
+            StartAfter=self.bucketKey
+        )["Contents"]
+        
+        return response
+
+
+    def __releaseS3Lock(self):
+        thisLockfile = path.join(self.bucketKey, self.lockfileName)
+        if self.debugPrint:
+            print(f"Removing lock {thisLockfile}")
+        self.s3Resource.Object(self.bucketName, thisLockfile).delete()
 
 
     ###########################################################################
     # S3 utilities
     ###########################################################################
-    def __downloadIfS3(self):
-        # self.localS3Copy = tempfile.TemporaryDirectory(...)
-        #
-        # Download bucket contents and unpack in self.localS3Copy.name
-        #
+    def __downloadIfS3(self, results=False):
+        self.localS3Copy = tempfile.TemporaryDirectory(suffix="asv")
+
+        # If results isn't set, only download key files, else download key files and results
+        if results == False:
+            self.s3Resource.Object(self.bucketName, path.join(self.bucketKey, confFileExt))
+                .download_file(path.join(self.localS3Copy.name, confFileExt))
+            self.s3Resource.Object(self.bucketName, path.join(self.bucketKey, benchmarksFileExtt))
```

Comment:
```suggestion
            self.s3Resource.Object(self.bucketName, path.join(self.bucketKey, benchmarksFileExt))
```

looks like a typo here